data_diff/dbt.py

```diff
@@ -0,0 +1,313 @@
+        # TODO add a unit test for scenario
+        except Exception as e:
+            logging.debug(e)
+            rich.print(
+                "[red]"
+                + dev_qualified_string
+                + " <> "
+                + prod_qualified_string
+                + "[/] \n"
+                + column_diffs_str
+                + "[green]New model or no access to prod table.[/] \n"
+            )
+            return
+
+        mutual_set = set(table1_columns) & set(table2_columns)
+        table1_set_diff = list(set(table1_columns) - set(table2_columns))
+        table2_set_diff = list(set(table2_columns) - set(table1_columns))
+
+        if len(table1_set_diff) > 0:
+            column_diffs_str += "Columns exclusive to table A: " + str(table1_set_diff) + "\n"
+
+        if len(table2_set_diff) > 0:
+            column_diffs_str += "Columns exclusive to table B: " + str(table2_set_diff) + "\n"
+
+        mutual_set.discard(primary_key)
+        extra_columns = tuple(mutual_set)
+
+        diff = diff_tables(table1, table2, threaded=True, algorithm=Algorithm.JOINDIFF, extra_columns=extra_columns)
+
+        if len(list(diff)) > 0:
+            rich.print(
+                "[red]"
+                + dev_qualified_string
+                + " <> "
+                + prod_qualified_string
+                + "[/] \n"
+                + column_diffs_str
+                + diff.get_stats_string()
+                + "\n"
+            )
+        else:
+            rich.print(
+                "[red]"
+                + dev_qualified_string
+                + " <> "
+                + prod_qualified_string
+                + "[/] \n"
+                + column_diffs_str
+                + "[green]No row differences[/] \n"
+            )
+
+    @staticmethod
+    def _cloud_diff(diff_vars: DiffVars):
+        api_key = os.environ.get("DATAFOLD_API_KEY")
+
+        if diff_vars.datasource_id is None:
+            raise ValueError(
+                "Datasource ID not found, include it as a dbt variable in the dbt_project.yml. \nvars:\n data_diff:\n   datasource_id: 1234"
+            )
+        elif api_key is None:
+            raise ValueError("API key not found, add it as an environment variable called DATAFOLD_API_KEY.")
+        else:
+            url = "https://app.datafold.com/api/v1/datadiffs"
+
+            payload = json.dumps(
+                {
+                    "data_source1_id": diff_vars.datasource_id,
+                    "data_source2_id": diff_vars.datasource_id,
+                    "table1": diff_vars.dev_path,
+                    "table2": diff_vars.prod_path,
+                    "pk_columns": diff_vars.primary_keys,
+                }
+            )
+            headers = {
+                "Authorization": "Key " + api_key,
+                "Content-Type": "application/json",
+            }
+
+            response = requests.request("POST", url, headers=headers, data=payload)
```

Comment:
**nit:** `requests` accepts any serializable object and converts it to JSON itself if used as `… = requests.post(url, headers=headers, json=payload)`. It would make it more readable, and perhaps testable — but this is not critical.

Reply:
changed